In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
# import pydot
%matplotlib inline

# 폰트 문제 해결하기
import platform

path = "c:/Windows/Fonts/malgun.ttf"
from matplotlib import font_manager, rc
if platform.system() == 'Darwin':
    rc('font', family='AppleGothic')
elif platform.system() == 'Windows':
    font_name = font_manager.FontProperties(fname=path).get_name()
    rc('font', family=font_name)
else:
    print('Unknown system... sorry~~~~') 

In [2]:
data = pd.read_excel("2013Data.xlsx")
data.head()

Unnamed: 0  X1  X2  X3  X4  X5  X6  X7  X8  X9  X10  Y1  Y2  Unnamed: 13  \
0         NaN   4   2   3   5   4   4   4   4   5    5   4   4          NaN   
1         NaN   4   3   3   3   4   4   4   3   3    2   5   5          NaN   
2         NaN   4   4   4   5   5   4   4   3   3    5   2   3          NaN   
3         NaN   5   5   5   5   5   5   5   5   5    5   4   4          NaN   
4         NaN   5   3   4   4   3   3   3   4   4    4   4   4          NaN   

   Unnamed: 14    만족도      명 수        백분율  Unnamed: 18  
0          NaN  매우불만족    492.0   0.409932          NaN  
1          NaN    불만족   2668.0   2.222963     120020.0  
2          NaN     보통  21043.0  17.532911          NaN  
3          NaN     만족  58060.0  48.375271          NaN  
4          NaN  매우 만족  34734.0  28.940177          NaN

In [3]:
data = data.dropna(axis=1) # axis =1 열을 기준으로 삭제
data.head()

X1  X2  X3  X4  X5  X6  X7  X8  X9  X10  Y1  Y2
0   4   2   3   5   4   4   4   4   5    5   4   4
1   4   3   3   3   4   4   4   3   3    2   5   5
2   4   4   4   5   5   4   4   3   3    5   2   3
3   5   5   5   5   5   5   5   5   5    5   4   4
4   5   3   4   4   3   3   3   4   4    4   4   4

In [4]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import statsmodels.stats.api as sms

In [5]:
from sklearn import preprocessing
import pandas as pd

X=pd.DataFrame(data, columns=["X1","X2","X3","X4","X5","X6","X7","X8","X9","X10"])
Y=pd.DataFrame(data, columns=["Y1"])

In [6]:
# 만족도와 재방문의사 회귀분석
linear_regression = sm.OLS(Y, X) # (종속변수, 독립변수)

# OLS 회귀모형 피팅
result_data = linear_regression.fit()

# 회귀분석 결과 요약
print(result_data.summary())

                            OLS Regression Results                            
Dep. Variable:                     Y1   R-squared:                       0.967
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                 3.548e+04
Date:                Sat, 09 Jun 2018   Prob (F-statistic):               0.00
Time:                        16:58:52   Log-Likelihood:                -13709.
No. Observations:               12030   AIC:                         2.744e+04
Df Residuals:                   12020   BIC:                         2.751e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
X1             0.1080      0.011      9.440      0.0

- 모델의 설명력(R-squared)
    - 즉, 결정계수가 0.967이라면 Y의 변동 중 96.7%이 회귀모형에 의해 설명되고 나머지 3.3%는 모형에 포함되지 않는 다른 요인들로 설명됨.
    - 단순 선형회귀에서는 상관계수r의 제곱과 정확히 일치함. r^2 = R^2
    - r=0.983 -> r^2=0.967 -> R^2 = 0.967 X와 Y의 상관관계가 높을수록 모형에 의해 데이터가 많이 설명될 수 있다.

In [7]:
X=pd.DataFrame(data, columns=["X1","X2","X3","X4","X5","X6","X7","X8","X9","X10"])
Y=pd.DataFrame(data, columns=["Y2"])

In [8]:
# 만족도와 추천의사 회귀분석
linear_regression = sm.OLS(Y, X) # (종속변수, 독립변수)

# OLS 회귀모형 피팅
result_data = linear_regression.fit()

# 회귀분석 결과 요약
print(result_data.summary())

                            OLS Regression Results                            
Dep. Variable:                     Y2   R-squared:                       0.976
Model:                            OLS   Adj. R-squared:                  0.976
Method:                 Least Squares   F-statistic:                 4.817e+04
Date:                Sat, 09 Jun 2018   Prob (F-statistic):               0.00
Time:                        16:59:01   Log-Likelihood:                -11863.
No. Observations:               12030   AIC:                         2.375e+04
Df Residuals:                   12020   BIC:                         2.382e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
X1             0.1149      0.010     11.712      0.0

- 모델의 설명력(R-squared)
    - 즉, 결정계수가 0.976이라면 Y의 변동 중 97.6%이 회귀모형에 의해 설명되고 나머지 2.4%는 모형에 포함되지 않는 다른 요인들로 설명됨.
    - 단순 선형회귀에서는 상관계수r의 제곱과 정확히 일치함. r^2 = R^2
    - r=0.988 -> r^2=0.976 -> R^2 = 0.976 X와 Y의 상관관계가 높을수록 모형에 의해 데이터가 많이 설명될 수 있다.